### DISCUSSION TITLE
2.0 Hybrid Mode Early Preview & Feedback Thread

### DISCUSSION THREAD

**Discussion by johnsoncodehk** [vuejs/language-tools#3789] [General]
In the next 2.0 release, we'll implement typescript support through TS plugin integration.

Since Vue no longer needs to run its own TS language server, this will result in a significant reduction in memory usage and an increase in performance. (This was previously done via takeover mode, but is no longer necessary.)

The basic language support is there now, but there is still a lot of additional functionality that needs to be re-implemented through the TS plugin, which won't be done anytime soon.

For those of you who wish to participate in early testing, I will keep posting new versions of the TS plugin integration in this post until the #3788 merge.

There are still a lot of issues with earlier versions, and for the obvious ones it should be fixed soon, so if you find something that hasn't been fixed, then I may have missed it, so please raise it in this thread instead of opening an issue.

### Early Preview Download

After downloading and unzipping it, you can install it via the VSCode "Extensions: Install from VSIX..." command.

- v5 (2023-12-15): [volar-2.0.0-94c4653e.5.vsix.zip](https://github.com/vuejs/language-tools/files/13683672/volar-2.0.0-94c4653e.5.vsix.zip)
- v4 (2023-12-11): [volar-2.0.0-f7c91202.4.vsix.zip](https://github.com/vuejs/language-tools/files/13629326/volar-2.0.0-f7c91202.4.vsix.zip)
- v3 (2023-12-10): [volar-2.0.0-17cbf2dc.3.vsix.zip](https://github.com/vuejs/language-tools/files/13625798/volar-2.0.0-17cbf2dc.3.vsix.zip)
- v2 (2023-12-08): [volar-2.0.0-d6afc068.2.vsix.zip](https://github.com/vuejs/language-tools/files/13607490/volar-2.0.0-d6afc068.2.vsix.zip)
- v1 (2023-12-08): [volar-2.0.0-df8b000a.1.vsix.zip](https://github.com/vuejs/language-tools/files/13605042/volar-2.0.0-df8b000a.1.vsix.zip)

**Comment by haoziqaq** [comment]
Reduced memory usage by 300m!

**Comment by they-cloned-me** [comment]
🔥 Thank you Johnson

**Comment by troy351** [comment]
Could you provide any tips for why Hybrid Mode not possible before?

**Comment by warmthsea** [comment]
🔥

**Comment by warmthsea** [comment]
It would be great if .vue file could be automatically repaired and imported like .ts file.

![screenshots](https://github.com/vuejs/language-tools/assets/45450994/ef631b49-1389-48cc-990a-d170699895e1)

**Reply by so1ve** [reply]
It should work in 2.0

**Comment by zouhangwithsweet** [comment]
note: `<template v-for> key should be placed on the <template> tag.`
<img width="572" alt="image" src="https://github.com/vuejs/language-tools/assets/21095710/f4d36b6e-2e94-41c2-a031-ad1e25964c06">

**Comment by rchl** [comment]
> After downloading and unzipping it, you can install it via the VSCode "Extensions: Install from VSIX..." command.

Maybe it's a bit early for that but at some point maybe you could also distribute it as a plain plugin so that it would be easier for other editors than VSCode to utilize it.

I guess it could be distributed as a separate npm package.

**Comment by rchl** [comment]
Do you plan to support `experimentalAdditionalLanguageModules` in this approach?

**Comment by rchl** [comment]
Just a heads up that typescript-language-features handling of plugins might be a bit lacking in case of workspace configuration handling. For example due to [logic like here](https://github.com/microsoft/vscode/blob/82ece597ee428f9fdb6e45698ba273cadb5d5202/extensions/typescript-language-features/src/languageFeatures/fileConfigurationManager.ts#L168-L171), it will apply `javascript.*` workspace configuration for .vue files.

**Comment by Loongphy** [comment]
If you got errors like `Cannot find name '__VLS_ctx'.`, add `"types": ["vite/client"]` in the `tsconfig.json`.

![image](https://github.com/vuejs/language-tools/assets/42089082/c112a2e2-5810-4b6c-87a5-8547aeb54404)

**Comment by skywalker512** [comment]
1. There seems to be an issue with automatic import/complete of TypeScript in the default Vue TS template created using Vite.

    https://github.com/vuejs/language-tools/assets/12960906/5f3db15d-7f15-404f-a2b3-7b2938e2ff72
    
    https://github.com/vuejs/language-tools/assets/12960906/f3cccc4e-51fd-4f16-91bd-69637ff66894

2. It seems that memory usage is greater than take over mode.

    - use 2.0 Hybrid Mode
    
    ![image](https://github.com/vuejs/language-tools/assets/12960906/dfc754ce-03fc-4637-9b27-738830a7a0a7)
    
    - use 1.8.25 take over mode
    
    ![image](https://github.com/vuejs/language-tools/assets/12960906/2c925b38-5967-471f-9331-0c014313053d)
    
    - use 1.8.25 and ts plugin
    
    ![image](https://github.com/vuejs/language-tools/assets/12960906/5630606a-4391-4bdf-a999-9c01e78bb7a4)

**Reply by johnsoncodehk** [reply]
The first video problem is from upstream and it will fix by https://github.com/microsoft/vscode/pull/200945.
The second video problem and memory usage regression should be fixed in v5 by https://github.com/volarjs/volar.js/commit/a0d92954f758dfe62e0c52c9b0a0d4cc06738c74 and #3803, please take a try. Thanks!

**Comment by yyx990803** [comment]
I noticed in 2.0 the syntax highlighting seems to be a bit off - components are not getting a different color from native elements.

**Reply by so1ve** [reply]
This could be caused by semantic tokens?

**Reply by Tenrys** [reply]
I couldn't get any auto-completion working on them as well. Auto-importing them made them auto-import and auto-complete as `MyComponentVue`.

**Comment by Tenrys** [comment]
I noticed that the language server was freaking out, constantly finding errors and then cleaning them out, refreshing constantly.

**Reply by Tenrys** [reply]
I should mention that this happens with this feature enabled: 
![image](https://github.com/vuejs/language-tools/assets/3979239/0e6367f6-1505-47e2-9e7b-1a4bca61ffaf)

**Comment by Cheerwhy** [comment]
component‘s highlight isn't working in template

**Comment by Eazash** [comment]
After installing v5, ts language server crashes and this is the output

<img width="1333" alt="Screenshot 2023-12-26 at 9 53 17 AM" src="https://github.com/vuejs/language-tools/assets/56728970/03baa78e-6056-4605-a15e-6a46601beb35">

**Reply by productdevbook** [reply]
this happens to me a lot too. but I don't know where it originated from.

**Reply by Eazash** [reply]
v4 doesn't crash the Typescript language server but it does have the same message 
<img width="1350" alt="image" src="https://github.com/vuejs/language-tools/assets/56728970/2e257142-3e22-4794-b37f-9d8b3eecf313">

**Comment by rennzhang** [comment]
Hi, also I see yesterday's update has been closed #3329 ，But problems(https://github.com/vuejs/language-tools/issues/3329#issuecomment-1859693571) remain

_Maybe I should open a new issue?_

### You can follow these steps to repeat
I have provided the following minimal reproducible cases, hoping to provide some help。

repo：https://github.com/rennzhang/vue3-ts-error-report

> This project uses the `unplugin-auto-import` plug-in.

1. Create a new folder under views, Like "aaa" 
2. Create `index.vue` under "aaa"
3. Enter the following code & `npm run dev`

```vue
<template>
  <div></div>
</template>

<script lang="ts" setup>
let a = ref(1);
</script>

<style scoped lang="less"></style>
```

The following error will occur

```
Name not found “ref”。ts(2304)
```
<img width="1440" alt="image" src="https://github.com/vuejs/language-tools/assets/50518861/8e22fa88-dce6-48b7-9afc-47ae336e1ba6">

**Reply by sadeghbarati** [reply]
@rennzhang 

Just tested your repository (with VSCode Profiles) with Volar `v1.8.27` and `volar-2.0.0-94c4653e.5.vsix` both of them working well

Also check this boilerplate seems a better starter project

https://github.com/dishait/tov-template

---

<details><summary>With Volar <=1.8.27</summary>
<p>

#### `extensions`

- Vue Language Features (Volar)

- TypeScript Vue Plugin (Volar)

#### `configurations`

- No need to add `shims-vue.d.ts` when using Volar

- Disabled VSCode TypeScript built-in (Takeover mode)👇 

  <details><summary>Details</summary>
  <p>
  
  ![image](https://github.com/vuejs/language-tools/assets/17789047/91aa4b52-dac5-4499-a322-0d70ca1672a8)
  
  
  </p>
  </details> 
  
- Add `vueTemplate: true` to your Vite `AutoImport` plugin
  
  </p>
  </details> 
  
  
---
  
<details><summary>With volar-2.0.0-94c4653e.5.vsix (>=2)</summary>
<p>

#### `extensions`

- Vue Language Features (Volar) (Manually installed version of volar with that .vsix file)

#### `configurations`

- No need to add `shims-vue.d.ts` when using Volar

- No need to disable TypeScript built-in in Volar 2? (No Takeovermode)

- Add `vueTemplate: true` to your Vite `AutoImport` plugin

![image](https://github.com/vuejs/language-tools/assets/17789047/42a644f6-6191-497f-86d4-fc3c1cc01504)


</p>
</details>

**Reply by rennzhang** [reply]
Thank you very much, but after I adjust it according to your steps, there is still the same problem when using Volar <= 1.8.27.

But switching to Volar-2.0.0-94C4653E.5.VSIX (> = 2) is okay. Since this is so, I plan to use V2 directly in advance.

**Comment by rodrigocfd** [comment]
I have [project-wide error reporting](https://github.com/microsoft/vscode/issues/13953#issuecomment-583179315) enabled for TypeScript.

In the example below, I change the name of a function, thus causing an error in the files that import it. When an error happens, the affected files should go red... but instead, the error report keeps going on and off:

![bizarre-flash](https://github.com/vuejs/language-tools/assets/1283868/02fd7372-a040-40c9-a4b2-adedb4db126b)

I'm using:
* Volar v5 (2023-12-15);
* VSCode v1.85.2;
* Linux Mint v21.3.

**Comment by Eazash** [comment]
I'm seeing an issue where selecting a component from autocomplete adds an import statement in nuxt projects. Also, if there's no script setup section, the import statement simply goes to the top of the file. 


https://github.com/vuejs/language-tools/assets/56728970/762a09af-0b70-4579-b000-f8c542bbd6b1

I'm using:
- Volar v5 (2023-12-15)
- Vscode insiders, January 2024 (version 1.86)
- Vue 3.43
- Nuxt 3.9.1

Not attaching a reproduction repo as any starter nuxt app has the same issue

**Comment by jaskang** [comment]
Do we have the latest preview version available? plz 🙏

**Comment by baixiaoyu2997** [comment]
test with v5
the issue still: https://github.com/vuejs/language-tools/issues/3390
